テーブル設計

3so opened the pull request:
#WHAT
Furimaアプリに必要なエンティティの洗い出しとER図の作成。
テーブル設計をREADMEへ記述。

ER図は以下の内容となっております。
https://gyazo.com/e9b91e6ed0666e8c8a47f997c80e9c32

#WHY
実際にテーブルを作成する前に、必要な要素を予め整理しておく必要があるため。

3so committed:
Robocopの導入

3so committed:
rails_12factorの実装

3so committed:
Revert "rails_12factorの実装"

This reverts commit 1a8432292d875cd61021f4ff6a92890bde6af28a.

3so committed:
Revert "Revert "rails_12factorの実装""

This reverts commit eb4f4709fd48cfdf9891e251e8e94b2178f87a77.

3so committed:
Basic認証を導入

3so committed:
テーブル設計

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,58 @@
+| first_name          | string  | null: false |
+| family_name_reading | string  | null: false |
+| first_name_reading  | string  | null: false |
+| birth_year          | integer | null: false |
+| birth_month         | integer | null: false |
+| birth_day           | integer | null: false |
```
誕生日カラムはひとまとめにしてdate型にしましょう！

3so replied:
誕生日カラムをbirthdayカラムにまとめ、date型に変更しました。

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,58 @@
-* Deployment instructions
-
-* ...
+# テーブル設計
```
商品は購入した後取引を行います、その際使用するのはuser_idとitem_idです！
それを管理するテーブルが必要ですので、設計してみましょう！

3so replied:
取引情報を管理するテーブルとしてpurchasesテーブルを設計し、user_idカラム・item_idカラムをそれぞれreferences型で設定いたしました。

syuji-yamamoto commented on README.md:
```diff
@@ -1,24 +1,58 @@
+| house_number  | string     | null: false                    |
+| building_name | string     |                                |
+| phone_number  | integer    | null: false                    |
+| user          | references | null: false, foreign_key: true |
+| item          | references | null: false, foreign_key: true |
```
今回の設計では「購入管理テーブルに配送先が紐づく」という実装を行なっていきたいので、商品やユーザーの外部キーカラムではなく、購入管理テーブルの外部キーカラムを用意してあげましょう。

3so replied:
取引情報と配送先住所をpurchasesテーブルにまとめて記述してしまっていたため、取引情報をpurchasesテーブルに、配送先住所をshipping_addressesテーブルに記載するよう変更いたしました。
shipping_addressesテーブルにpurchases_idカラムをreference型で設定いたしました。

syuji-yamamoto reviewed (commented):
レビュー依頼ありがとうございます！
綺麗に設計できていて素晴らしいですね。
コメントしましたので、ご確認ください！

3so committed:
テーブル設計の変更

3so commented:
コメントいただき、ありがとうございます。
各コメントに対し、変更点を返信させていただきました。

また、ER図を以下のように変更しております。
https://gyazo.com/98ef492f481bdfab06606b875a60aded

お手数をおかけいたしますが、ご確認よろしくお願いいたします。

i-date commented on README.md:
```diff
@@ -1,24 +1,67 @@
 
-* System dependencies
+- has_many :items
+- has_many :Purchases
```
19行目にて、`:Purchases`と大文字の`P`になっておりますので小文字に修正しましょう。

3so replied:
小文字に変更いたしました。

i-date commented on README.md:
```diff
@@ -1,24 +1,67 @@
+| ---------------------- | ---------- | ------------------------------ |
+| name                   | string     | null: false                    |
+| description            | text       | null: false                    |
+| condition              | integer    | null: false                    |
+| category               | integer    | null: false                    |
+| shipping_charges_payer | integer    | null: false                    |
+| shipping_location      | integer    | null: false                    |
+| lead_time              | integer    | null: false                    |
```
こちら後々のエラーの原因になってしまう可能性があるので指摘させていただきます。
active_hashで導入するこれらのカラム名は語尾に「`_id`」をつけましょう。
例：`condition_id`

3so replied:
active_hashをで導入する予定のカラム名について、それぞれ末尾に「_id」をつけました。

i-date commented on README.md:
```diff
@@ -1,24 +1,67 @@
+
+| Column        | Type       | Options                        |
+| ------------- | ---------- | ------------------------------ |
+| postal_code   | integer    | null: false                    |
```
57行目の郵便番号を管理するカラムですが、商品購入機能の実装条件にある「郵便番号にはハイフンが必要であること（123-4567となる）」も考慮し、integer型ではなくてstring型にしましょう。

3so replied:
実装条件に合わせ、string型に変更いたしました。

i-date commented on README.md:
```diff
@@ -1,24 +1,67 @@
+| Column        | Type       | Options                        |
+| ------------- | ---------- | ------------------------------ |
+| postal_code   | integer    | null: false                    |
+| prefecture    | integer    | null: false                    |
```
こちら後々のエラーの原因になってしまう可能性があるので指摘させていただきます。
58行目のactive_hashで導入する都道府県を管理するカラムのカラム名は語尾に「_id」をつけましょう。

3so replied:
prefectureカラムのカラム名末尾に「_id」をつけました。

i-date commented on README.md:
```diff
@@ -1,24 +1,67 @@
+| city          | string     | null: false                    |
+| house_number  | string     | null: false                    |
+| building_name | string     |                                |
+| phone_number  | integer    | null: false                    |
```
62行目の電話番号を管理するカラムですが、電話番号は数字の文字列として実装しておきたい（数値で管理すると頭の0を補完する必要がある）ため、integer型ではなくてstring型にしましょう。

3so replied:
頭の0を補完する必要がないよう、phone_numberカラムをstring型に変更しました。

i-date reviewed (commented):
修正ありがとうございます！
追加コメントいたしましたので、ご確認をお願いいたします。

3so committed:
テーブル設計の変更

・テーブル名の文頭が大文字になっていた箇所を小文字に変更
・active_hashを利用して実装するカラムのカラム名末尾に「_id」を追加
・実装上不都合が生じるため、postal_codeカラムとphone_numberカラムをstring型に変更

3so commented:
コメントいただき、ありがとうございます。
各コメントに対し、修正内容を返信させていただきました。

また、ER図を以下のように変更いたしました。
https://gyazo.com/ce546ca8772d59cf4040f65d742fd69f

以上、お手数をおかけいたしますが、ご確認よろしくお願いいたします。

i-date reviewed (approved):
修正とER図ありがとうございます！
お疲れさまでした、LGTMとさせていただきます。
以降の実装も引き続きこの調子で進めていきましょう。応援しております！
